src/users/user.service.spec.ts

```diff
@@ -0,0 +1,146 @@
+const crypto = require("crypto");
+import { Test, TestingModule } from "@nestjs/testing";
+import { DBAccessModule } from "@wiredcraft/dbaccess/dbaccess.module";
+import { PrismaService } from "@wiredcraft/dbaccess/prisma.service";
+import { createPrismaMockService } from "@wiredcraft/mocks/prisma.mock.service";
+import UserDto from "./dto/user.dto";
+import { UserService } from "./user.service";
+import { ConflictException } from "@nestjs/common";
+import { excludedFields } from "@wiredcraft/utils/comm.util";
+
+//create random date for dob
+function randomDate(start = new Date(2012, 0, 1), end = new Date()) {
+  return new Date(
+    start.getTime() + Math.random() * (end.getTime() - start.getTime())
+  );
+}
+
+//helper to compare two user
+function userEquals(u1: UserDto, u2: UserDto) {
+  expect(u1.name).toBe(u2.name);
+  expect(u1.email).toBe(u2.email);
+  expect(u1.dob).toStrictEqual(u2.dob);
+  expect(u1.address).toBe(u2.address);
+  expect(u1.description).toBe(u2.description);
+}
+
+//helper to create userDto
+function createUserDto() {
+  const randomStr = crypto.randomUUID();
+  return {
+    name: randomStr,
+    dob: randomDate(),
+    email: `${randomStr}@wiredcard.com`,
+    address: `${randomStr} address`,
+    description: `hi this is ${randomStr}`,
+  } as UserDto;
+}
+describe("UserService", () => {
+  const dbName = "user_unit_test";
+  let service: UserService;
+  let prisma: PrismaService;
+  let mongoReplSet;
+  beforeAll(async () => {
+    mongoReplSet = await createPrismaMockService(dbName);
+    await mongoReplSet.waitUntilRunning();
+    prisma = new PrismaService({
+      datasources: {
+        db: {
+          url: mongoReplSet.getUri(dbName),
+        },
+      },
+    });
+    const module: TestingModule = await Test.createTestingModule({
+      imports: [DBAccessModule],
+      providers: [UserService],
+    })
+      .overrideProvider(PrismaService)
+      .useValue(prisma)
+      .compile();
+
+    service = module.get<UserService>(UserService);
+  }, 1000 * 30);
+  beforeEach(async () => {
+    await prisma.user.deleteMany();
+    await prisma.$disconnect();
+    await prisma.$connect();
+  });
+  afterAll(async () => {
+    await prisma.$disconnect();
+    await mongoReplSet.stop();
+  }, 1000 * 30);
+  describe("create", () => {
+    it("should create user success", async () => {
+      const newUser = createUserDto();
+
+      const createdUser = await service.create(newUser);
+      expect(createdUser.id).toBeDefined();
+      userEquals(newUser, createdUser);
+    });
+    it("should create user failed, duplicated email address", async () => {
```

Comment:
in which test setup step is index created?
in deployment process of QA/Staging/Production, how to create required index and how to know index is created successfully?

Reply:
there is no index added in this project. I think it is not good practices add index at beginning of the new project, unless it has fairly a mount of preloaded data in db otherwise Create index in empty db may cause slowness insertion.

To create required index usually it should be done by db migration scripts instead add indexes directly on database, and we can use post deployment script check if index created properly. Or use db monitor tool, if it is not first release we also need do the regression test make sure everything works fine.

Reply:
you declared user's `email` field to be unique, even without index, the app still checks in slower way to find if there is another user with same `email` when insert / update.
would the index make insertion faster in this case?

Reply:
you declared user's email field to be unique, even without index, the app still checks in slower way to find if there is another user with same email when insert / update.
would the index make insertion faster in this case?
> lol I think I misunderstood your question previously and also in my code I did something wrong, I didnt mean to add unique index, I'm trying to add comment to email field. it should be //unique instead of @unique.
for the database index it really depends on cases. so I didn't try add any in this project.